### DISCUSSION TITLE
PLEASE HELP!!! How to create service available globally within the application not just nitro or nuxtapp?

### DISCUSSION THREAD

**Discussion by weloin-sukanta** [nuxt/nuxt#29761] [Questions] [answered]
I tried several options but couldn't find appropriate way of registering a singleton service which itself registers several functionality. For example, I've a Service called "Workflow" and I want to register activities inside this workflow, something like:

```
import {Activity1, Activity2, ....} from "/path/to/activities".
...
workflowService.register([Activity1, Activity2, ....]);
```

I need to run this only once, as it is expensive to call at every api call or cron jobs, or any other background stuffs. Remember I need this to only work in the server side, client side is not needed. I tried registering it with `plugins/<plugin>.server.ts`, and `server/plugins/<plugin>.ts` none of them works, as the context is completely different when using the service inside a cron job.

I couldn't understand the underlying concept here, what is the best way to do this?

PLEASE HELP!!!

**Comment by pi0** [comment]
Can you please explain your use case in a simpler/non-generic way? 

What are those activities? Normally you can use [Nitro scheduled tasks](https://nitro.build/guide/tasks#scheduled-tasks) but if it is more clear to me what you want to exactly do i might help better.

**Comment by weloin-sukanta** [comment]
Hi @pi0, thanks for your reply also sorry about the confusion, actually the implementation doesn't matter. Let me clarify what I'm trying to achieve.

I want to create a system where I can register several classes (here activities are just classes). Lets say I've 100s of activities. for example:

- Transform Data
- Send Email
- Send SMS
- Call Webhook
and so on...

Here is an example of activity class:

```ts
class SendEmailActivity extends AbstractActivity{
	override key(): string{
		return 'send_email';
	},
	override run(config): any {
		// Execute the activity.
	}
}

```

An workflow is basically some combination of some of the activities, with configuration data for each activity. Each Activity must be registered with a unique key, (somewhat similar to modules in nuxt). I may call the service to register it from the API or at boot time, something like this:

```ts
WorkflowService.register(SendEmailActivity)
WorkflowService.register(SendSMSActivity)
...
...
```

`WorkflowService` contains all the registered activity classes in some variables, so when it needs to instantiate, it just matches with the `key()` and instantiates it for 


When I want to save a workflow, it saves in some kind of this format (removing obvious details):

```ts
const workflow = {
 id: '......',
 activities: [
  {key: send_email, options: {...}},
  {key: call_webhook, options: {...}},
  ...
  ...
 ]
}
```

Now the above is the format that is now loaded in memory from database, lets say in an API handler or cron-job handler or scheduled tasks (anywhere in the server side). I tried register with plugin based system, so that other plugins could register their own Activities.

`WorkflowService` contains all the registered activity classes in some variables, so when it needs to instantiate, it just matches with the `key()` and instantiates it for execution. Now, when I register it in plugin, the API handler shows nothing is registered there. Also, when calling it from cron jobs, it says empty. So, I had to register all the activities right at that context when using it: in the API handler, cron jobs, etc, all separately, which removes the capacity of adding other activities from plugins.

By the way, I'm using https://github.com/hywax/nuxt-cron for cron jobs.

Sorry, for the long post to read, may be this is someting very obvious to you, but still haunting me. Please Help!

**Reply by pi0** [reply, marked as answer]
Since it is a global service, i think you can use a [Nitro plugin](https://nitro.build/guide/plugins) to register initial workflow services and in event handlers, you can access `nitroApp` with `useNitroApp()` if you want to update it (register new workflows, etc).

The same nitro instance is also accessible from [Nitro tasks](https://nitro.build/guide/tasks) (i highly recommend using this method instead of 3rd party module) so we can use it inside them to resolve the workflows.

I hope this is helpful. Feel free to also reach me out via Discord (pi0) if you have more specific questions.

**Reply by weloin-sukanta** [reply]
Too Good, this works like a charm. Thanks @pi0.

**Comment by weloin-sukanta** [comment]
Thanks for the quick reply @pi0. Let me test and will get back.
